tsdb/engine/tsm1/ring.go

```diff
@@ -0,0 +1,289 @@
+	// such that the first eight bits of a given hash will map directly to one
+	// of the N partitions.
+	for i := 0; i < len(r.continuum); i++ {
+		if (i == 0 || i%(256/n) == 0) && len(r.partitions) < n {
+			r.partitions = append(r.partitions, &partition{
+				store:          make(map[string]*entry),
+				entrySizeHints: make(map[uint64]int),
+			})
+		}
+		r.continuum[i] = r.partitions[len(r.partitions)-1]
+	}
+	return &r, nil
+}
+
+// reset resets the ring so it can be reused. Before removing references to entries
+// within each partition it gathers sizing information to provide hints when
+// reallocating entries in partition maps.
+//
+// reset is not safe for use by multiple goroutines.
+func (r *ring) reset() {
+	for _, partition := range r.partitions {
+		partition.reset()
+	}
+	r.keysHint = 0
+}
+
+// getPartition retrieves the hash ring partition associated with the provided
+// key.
+func (r *ring) getPartition(key string) *partition {
+	return r.continuum[int(uint8(xxhash.Sum64([]byte(key))))]
+}
+
+// entry returns the entry for the given key.
+// entry is safe for use by multiple goroutines.
+func (r *ring) entry(key string) (*entry, bool) {
+	return r.getPartition(key).entry(key)
+}
+
+// write writes values to the entry in the ring's partition associated with key.
+// If no entry exists for the key then one will be created.
+// write is safe for use by multiple goroutines.
+func (r *ring) write(key string, values Values) error {
+	return r.getPartition(key).write(key, values)
+}
+
+// add adds an entry to the ring.
+func (r *ring) add(key string, entry *entry) {
+	r.getPartition(key).add(key, entry)
+	atomic.AddInt64(&r.keysHint, 1)
+}
+
+// remove deletes the entry for the given key.
+// remove is safe for use by multiple goroutines.
+func (r *ring) remove(key string) {
+	r.getPartition(key).remove(key)
+	if r.keysHint > 0 {
+		atomic.AddInt64(&r.keysHint, -1)
+	}
+}
+
+// keys returns all the keys from all partitions in the hash ring. The returned
+// keys will be in order if sorted is true.
+func (r *ring) keys(sorted bool) []string {
+	keys := make([]string, 0, atomic.LoadInt64(&r.keysHint))
+	for _, p := range r.partitions {
+		keys = append(keys, p.keys()...)
+	}
+
+	if sorted {
+		sort.Strings(keys)
+	}
+	return keys
+}
+
+// apply applies the provided function to every entry in the ring under a read
+// lock using a separate goroutine for each partition. The provided function
+// will be called with each key and the corresponding entry. The first error
+// encountered will be returned, if any. apply is safe for use by multiple
+// goroutines.
+func (r *ring) apply(f func(string, *entry) error) error {
+
+	var (
+		wg  sync.WaitGroup
+		res = make(chan error, len(r.partitions))
+	)
+
+	for _, p := range r.partitions {
+		wg.Add(1)
+
+		go func(p *partition) {
+			defer wg.Done()
+
+			p.mu.RLock()
+			for k, e := range p.store {
+				if err := f(k, e); err != nil {
```

Comment:
When I got a lot of timeouts under high load, this was one of the spots that showed up where many goroutines where busy.  The issue was a that each `entry` was being sorted which was expensive and there were millions of them which I think caused CPU starvation for the waiting write HTTP goroutines.